### PR TITLE
Reject requests that have duplicate Content-Type / Host headers

### LINKED ISSUE #3366: Duplicate Content-Type Header Parsing Issue
**Version**
bacbf8a

**Platform**
Ubuntu 11.4.0-1ubuntu1~22.04

**Description**
Hello, I may have found a bug in gunicorn's parsing of HTTP requests with duplicate `Content-Type` header.
RFC 9110 says this:
> Although `Content-Type` is defined as a singleton field, it is sometimes incorrectly generated multiple times, resulting in a combined field value that appears to be a list. Recipients often attempt to handle this error by using the last syntactically valid member of the list, leading to potential interoperability and security issues if different implementations have different error handling behaviors.

Gunicorn does not reject such requests. For example:
```
POST / HTTP/1.1\r\n
Host: victim.com\r\n
Content-Type: text/plain\r\n
Content-Type: application/json\r\n
\r\n
```
```
$ echo -ne "POST / HTTP/1.1\r\nHost: victim.com\r\nContent-Type: text/plain\r\nContent-Type: application/json\r\n\r\n" | nc 172.18.0.7 80
HTTP/1.1 200 OK
Server: gunicorn
Date: Thu, 20 Mar 2025 15:42:12 GMT
Connection: keep-alive
Content-type: application/json
Content-Length: 161

{"headers":[["SE9TVA==","dmljdGltLmNvbQ=="],["Q09OVEVOVF9UWVBF","YXBwbGljYXRpb24vanNvbg=="]],"body":"","version":"SFRUUC8xLjE=","uri":"Lw==","method":"UE9TVA=="}
```
> Recipients often attempt to handle this error by using the last syntactically valid member of the list, leading to potential interoperability and security issues if different implementations have different error handling behaviors.

If different implementations have different error handling behaviors, there may be potential interoperability and security issues. So I might suggest rejecting such requests.

### PR DESCRIPTION
**Pull request by Cycloctane**
Duplicate Content-Type and Host headers in requests are not allowed in RFC9110 and can cause security concerns. This PR makes `parse_headers()` raise `InvalidHeader` on these malformed requests.

Fixes #3366

### CODE REVIEW

**Review by pajod — CHANGES_REQUESTED** [review]
This swaps mapping/merging order. This results in meaningful header content change for accepted requests while only justifying the decision to reject some requests. Some test like `tests/requests/valid/040_compat.http` - but actually testing the edge case - would need to be added.

**Review by benoitc — CHANGES_REQUESTED** [review]
Thanks for your PR! Please check my comment. Also I think it should be an option. Some servers rely on multiple headers. Can you do the changes?

File: gunicorn/http/message.py [hunk 1]
```diff
@@ -119,7 +121,7 @@ def parse(self, unreader):
 
     def parse_headers(self, data, from_trailer=False):
         cfg = self.cfg
-        headers = []
+        headers = {}
```
[the remark this answers is not in the document]
Comment (Cycloctane) [reply gunicorn/http/message.py]: > this would break the header order. 

I thought this is not likely to happen because python dict preserves insertion order? Also no existing tests break because of this.

> Some servers rely on multiple headers

This change shouldn't affect headers other than Host and Content-Type (they will be merged anyway in wsgi.py). I'm going to add more tests and make sure other headers are handled exactly the same way as before. (as @pajod pointed out).

[the remark this answers is not in the document]
Comment (Cycloctane) [reply gunicorn/http/message.py]: I've reconsidered my approach. "Last seen" pattern is indeed cleaner for this PR. I've updated my changes. Thanks for your review :)

**Comment by Cycloctane** [comment]
@benoitc Sorry for the delay. It seems that the project was refactored. Should we add duplicated headers checks for the new http c parser to gunicorn_h1c? Then I would like to open a pr in that repository also.

**Comment by benoitc** [comment]
Thanks for this. Rejecting these is the right call, and the RFC 9110 5.3 framing
of it is what made the scope obvious.

Merged in #3691 with two changes.

The check moved from `parse_headers()` to `_apply_header_policy()`. `_parse_fast()`
never calls `parse_headers()`, so as written the fast parser would still have
accepted duplicates, and the corpus case you added would have failed on its
`fast` variant. The shared hook keeps both parsers in step. The ASGI parsers
needed their own handling on top: `H1CProtocol` rejects duplicate
`Content-Length` internally but accepts `Host` and `Content-Type`.

`Content-Length` came out of the singleton set. Duplicates are already rejected
in `set_body_reader()`, which raises with the request attached so the access log
keeps the entry; the earlier check would have preempted that.

Closing in favour of #3691. Your `041` corpus case is in it, plus a `042` for
duplicate Host.
